### PR TITLE
Full codegen for `log`, `log2`, and `log10`

### PR DESCRIPTION
**Pull request by miladm**
Full codegen for `log`, `log2`, and `log10`

---

Generated `LazyIr.h`

```
class Log : public XlaNode {
 public:
  static torch::lazy::OpKind ClassOpKind() {
    return torch::lazy::OpKind(at::aten::log);
  }
  Log(const torch_xla::XlaValue& self, std::vector<torch::lazy::Shape>&& shapes)

      : XlaNode(torch::lazy::OpKind(at::aten::log),
              {self}, std::move(shapes),
              [&]() { return LogOutputShape(self); },
              /* num_outputs */ 1,
              torch::lazy::MHash())
  {  }
  std::string ToString() const override {
    std::stringstream ss;
    ss << XlaNode::ToString();
    return ss.str();
  }
  bool CanBeReused(const torch_xla::XlaValue& self) const {
    return false;
    }
  torch_xla::XlaOpVector Lower(LoweringContext* loctx) const override;
};

class Log10 : public XlaNode {
 public:
  static torch::lazy::OpKind ClassOpKind() {
    return torch::lazy::OpKind(at::aten::log10);
  }

  Log10(const torch_xla::XlaValue& self, std::vector<torch::lazy::Shape>&& shapes)

      : XlaNode(torch::lazy::OpKind(at::aten::log10),
              {self}, std::move(shapes),
              [&]() { return Log10OutputShape(self); },
              /* num_outputs */ 1,
              torch::lazy::MHash())
  {  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << XlaNode::ToString();
    return ss.str();
  }
  bool CanBeReused(const torch_xla::XlaValue& self) const {
    return false;
    }
  torch_xla::XlaOpVector Lower(LoweringContext* loctx) const override;
};

class Log2 : public XlaNode {
 public:
  static torch::lazy::OpKind ClassOpKind() {
    return torch::lazy::OpKind(at::aten::log2);
  }

  Log2(const torch_xla::XlaValue& self, std::vector<torch::lazy::Shape>&& shapes)

      : XlaNode(torch::lazy::OpKind(at::aten::log2),
              {self}, std::move(shapes),
              [&]() { return Log2OutputShape(self); },
              /* num_outputs */ 1,
              torch::lazy::MHash()
  {  }
  std::string ToString() const override {
    std::stringstream ss;
    ss << XlaNode::ToString();

    return ss.str();
  }
  bool CanBeReused(const torch_xla::XlaValue& self) const {
    return false;
    }
  torch_xla::XlaOpVector Lower(LoweringContext* loctx) const override;
};
```
---

Generate `XLANativeFunctions.cpp`

```
    at::Tensor XLANativeFunctions::log(const at::Tensor & self) {
        XLA_FN_COUNTER("xla::");
        auto common_device = torch_xla::bridge::GetXlaDevice(self);
        TORCH_INTERNAL_ASSERT(common_device);
        torch_xla::XLATensorPtr lazy_self = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(self, *common_device);
        torch::lazy::NodePtr node = torch::lazy::ReuseNode<Log>(lazy_self->GetIrValue());
        if (!node) {
            auto out_meta = at::meta::log(self);
            std::vector<torch::lazy::Shape> shapes{
        torch::lazy::Shape(out_meta.scalar_type(), out_meta.sizes().vec())};
            TORCH_INTERNAL_ASSERT(shapes.size() == 1);
            if(torch::lazy::symbolicShapeEnabled()){
                std::vector<torch::jit::IValue> inputs = { self };
                char* schema_str = "aten::log(Tensor self) -> Tensor";
                applySymbolicShapesOnLT(schema_str, inputs, shapes);
            }
            node = torch::lazy::MakeNode<Log>(lazy_self->GetIrValue(), std::move(shapes));
            CacheNode(node);
        }
        auto result = torch_xla::bridge::AtenFromXlaTensor(
                torch_xla::XLATensor::Create(std::move(node), *common_device));
        return result;
    };


    at::Tensor XLANativeFunctions::log10(const at::Tensor & self) {
        XLA_FN_COUNTER("xla::");
        auto common_device = torch_xla::bridge::GetXlaDevice(self);
        TORCH_INTERNAL_ASSERT(common_device);
        torch_xla::XLATensorPtr lazy_self = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(self, *common_device);
        torch::lazy::NodePtr node = torch::lazy::ReuseNode<Log10>(lazy_self->GetIrValue());
        if (!node) {
            auto out_meta = at::meta::log10(self);
            std::vector<torch::lazy::Shape> shapes{
        torch::lazy::Shape(out_meta.scalar_type(), out_meta.sizes().vec())};
            TORCH_INTERNAL_ASSERT(shapes.size() == 1);
            if(torch::lazy::symbolicShapeEnabled()){
                std::vector<torch::jit::IValue> inputs = { self };
                char* schema_str = "aten::log10(Tensor self) -> Tensor";
                applySymbolicShapesOnLT(schema_str, inputs, shapes);
            }

            node = torch::lazy::MakeNode<Log10>(lazy_self->GetIrValue(), std::move(shapes));
            CacheNode(node);
        }
        auto result = torch_xla::bridge::AtenFromXlaTensor(
                torch_xla::XLATensor::Create(std::move(node), *common_device));
        return result;
    };


    at::Tensor XLANativeFunctions::log2(const at::Tensor & self) {
        XLA_FN_COUNTER("xla::");
        auto common_device = torch_xla::bridge::GetXlaDevice(self);
        TORCH_INTERNAL_ASSERT(common_device);
        torch_xla::XLATensorPtr lazy_self = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(self, *common_device);
        torch::lazy::NodePtr node = torch::lazy::ReuseNode<Log2>(lazy_self->GetIrValue());
        if (!node) {
            auto out_meta = at::meta::log2(self);
            std::vector<torch::lazy::Shape> shapes{
        torch::lazy::Shape(out_meta.scalar_type(), out_meta.sizes().vec())};
            TORCH_INTERNAL_ASSERT(shapes.size() == 1);
            if(torch::lazy::symbolicShapeEnabled()){
                std::vector<torch::jit::IValue> inputs = { self };
                char* schema_str = "aten::log2(Tensor self) -> Tensor";
                applySymbolicShapesOnLT(schema_str, inputs, shapes);
            }

            node = torch::lazy::MakeNode<Log2>(lazy_self->GetIrValue(), std::move(shapes));
            CacheNode(node);
        }

        auto result = torch_xla::bridge::AtenFromXlaTensor(
                torch_xla::XLATensor::Create(std::move(node), *common_device));
        return result;
    };
```

### CODE REVIEW

File: torch_xla/csrc/ops/ops_lower_fn.cpp [hunk 1]
```diff
@@ -41,6 +41,23 @@ torch_xla::XlaOpVector Atanh::Lower(LoweringContext* loctx) const {
   return ReturnOp(xla::Atanh(xla_input), loctx);
 }
 
+torch_xla::XlaOpVector Log::Lower(LoweringContext* loctx) const {
+  xla::XlaOp xla_input = loctx->GetOutputOp(operand(0));
+  return ReturnOp(xla::Log(xla_input), loctx);
+}
+
+torch_xla::XlaOpVector Log2::Lower(LoweringContext* loctx) const {
+  double base = 2.0;
```
Comment (JackCaoG) [inline torch_xla/csrc/ops/ops_lower_fn.cpp:50]: nit, `static const double`

File: torch_xla/csrc/ops/ops_lower_fn.cpp [hunk 2]
```diff
@@ -41,6 +41,23 @@ torch_xla::XlaOpVector Atanh::Lower(LoweringContext* loctx) const {
...
+  xla::XlaOp xla_input = loctx->GetOutputOp(operand(0));
+  return ReturnOp(BuildLogBase(xla_input, base), loctx);
+}
+
+torch_xla::XlaOpVector Log10::Lower(LoweringContext* loctx) const {
+  double base = 10.0;
```
Comment (JackCaoG) [inline torch_xla/csrc/ops/ops_lower_fn.cpp:56]: ditto.

File: torch_xla/csrc/tensor_methods.cpp [hunk 3]
```diff
@@ -1678,26 +1678,6 @@ XLATensor XLATensor::linspace(const at::Scalar& start, const at::Scalar& end,
       element_type);
 }
 
-XLATensor XLATensor::log(const XLATensor& input) {
-  // Here we explictly pass c10::nullopt as logical_element_type because
-  // otherwise result will inherit the input's logical_element_type. In the
-  // case of log(int) -> float, we want to derive the dtype from IR value
-  // instead of input's logical_element_type.
```
Comment (JackCaoG) [inline torch_xla/csrc/tensor_methods.cpp:1685]: hmm, I think this should be fine. Now with the codegen we should compute the `logical_element_type` from computed `torch::lazy::shape`

**Review by JackCaoG — APPROVED** [review]
minor nits, otherwise lgtm

**Review by JackCaoG — CHANGES_REQUESTED** [review]
Seem like test failure is real
```
RuntimeError: Error while lowering: [Float[20]] aten::log
XLA builder error: INVALID_ARGUMENT: Expected element type in shape to be floating or complex for log operation; got S64.:
```

File: torch_xla/csrc/tensor_methods.cpp (hunk 3 above)
Comment (miladm) [reply torch_xla/csrc/tensor_methods.cpp:1685]: I added type casting to the lowering stage. Letting CI run the test as local testing is broken at the moment.

**Review by wonjoolee95 — APPROVED** [review]
Nice, thanks! Will let the CI confirm the tests.

**Comment by wonjoolee95** [comment]
Synced with @miladm offline, I'll be taking these ops over in a new PR.

**Pull request closed without merging**